### ISSUE TITLE
react/no-array-index-key shouldn't apply to Maps

### ISSUE THREAD

**Issue by miXwui** [yannickcr/eslint-plugin-react#2821]
I ran into an odd scenario/edgecase with the rule `eslint(react/no-array-index-key)` which warns `Do not use Array index in keys`

When using `.foreach()` on a Map, the "array" index, or really, the Map key)
> can be any value (including functions, objects, or any primitive).
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map

In the example below, I'm using the Map's keys (which are strings) as the key for some React element:
```js
const filterValues = new Map([
    ['foo', {
      name: 'blah',
      type: 'hi',
    }],
    ['bar', {
      name: 'blargh',
      type: 'you',
    }],
    ['baz', {
      name: 'bloop',
      type: 'cool',
    }],
  ]);

....

const filters = [];

const createFilters = () => filterValues.forEach((value, key) => {
  filters.push(
    <div
      id={key}
      key={key} // eslint thinks it's an array index, which it technically is, but it's a string not an integer here
    />,
  );
});
```
I don't think applies to JS Maps?:
> In cases where the array is sorted or an element is added to the beginning of the array, the index will be changed even though the element representing that index may be the same.
https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-array-index-key.md

Granted Map keys aren't guaranteed to be unique, but neither is taking the values from an array of objects, like in the first "not considered warnings" example
```js
// this is okay
things.map((thing) => (
  <Hello key={thing.id} />
));
```
https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-array-index-key.md

Semi-related: https://github.com/yannickcr/eslint-plugin-react/issues/1049

Meanwhile, trying to use for...of results in this:
`eslint(no-restricted-syntax)`
`iterators/generators require regenerator-runtime, which is too heavyweight for this guide to allow them. Separately, loops should be avoided in favor of array iterations.`

Sooo I think the rule should ignore JS Maps *shrugs*

**Comment by ljharb** [comment]
The airbnb config indeed bans for..of.

However, in this case, you can do:
```jsx
const createFilters = () => Array.from(filterValues, ([key, value]) => {
})
```
which avoids needing the `filters` array at all.

**Comment by miXwui** [comment]
Oh hey yeah that works, and I think that should cover all cases I can think of (when dynamically creating elements).

Closing this issue, thank you @ljharb!
